File: packages/webstreamr/lib/webstreamr/extractor/hubdrive.dart
/// Port of webstreamr/src/extractor/HubDrive.ts
library;

import 'package:html/parser.dart' as html_parser;

import '../types.dart';
import '../utils/fetcher.dart';
import '../webstreamr_parse.dart';
import 'extractor.dart';
import 'hubcloud.dart';

class HubDrive extends Extractor {
  final HubCloud hubCloud;
  HubDrive(super.fetcher, this.hubCloud);

  @override
  String get id => 'hubdrive';
  @override
  String get label => 'HubDrive';
  @override
  Duration get ttl => const Duration(hours: 12);
  @override
  int? get cacheVersion => 1;

  @override
  bool supports(Context ctx, Uri url) => url.host.contains('hubdrive');

  Future<List<InternalUrlResult>> _fromHubCloud(
      Context ctx, Uri hubCloudUrl, Meta meta) async {
    final results = await hubCloud.extract(ctx, hubCloudUrl, meta);
    return results
        .map((r) => InternalUrlResult(
              url: r.url,
              format: r.format,
              isExternal: r.isExternal,
              ytId: r.ytId,
              error: r.error,
              label: r.label,
              meta: r.meta,
              requestHeaders: r.requestHeaders,
            ))
        .toList();
  }

  @override
  Future<List<InternalUrlResult>> extractInternal(
      Context ctx, Uri url, Meta meta) async {
    final headers = {'Referer': meta.referer ?? url.toString()};
    final html =
        await fetcher.text(ctx, url, FetcherRequestConfig(headers: headers));
    final next = tryRustNextUrl(id, html, url.toString());
    if (next != null) {
      return _fromHubCloud(ctx, Uri.parse(next), meta);
    }

    final doc = html_parser.parse(html);
    final hubCloudA = doc.querySelectorAll('a').firstWhere(
          (a) => a.text.contains('HubCloud'),
          orElse: () => html_parser.parseFragment('').nodes.isEmpty
              ? throw StateError('no anchors')
              : html_parser.parseFragment('<a></a>').children.first,
        );
    final hubCloudHref = hubCloudA.attributes['href'];
    if (hubCloudHref == null || hubCloudHref.isEmpty) return const [];
    return _fromHubCloud(ctx, Uri.parse(hubCloudHref), meta);
  }
}

File: crates/forja-webstreamr/src/extractors/hubdrive.rs
use crate::types::{ExtractResult, StreamFormat};
use regex::Regex;
use std::sync::LazyLock;

static HUBCLOUD_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r#"(?i)href=["']([^"']+)["'][^>]*>[\s\S]*?HubCloud"#).unwrap()
});

pub fn supports_host(host: &str) -> bool {
    host.contains("hubdrive")
}

pub fn extract_from_html(html: &str, _page_url: &str) -> Option<ExtractResult> {
    let href = HUBCLOUD_RE.captures(html)?.get(1)?.as_str();
    Some(ExtractResult {
        url: String::new(),
        format: StreamFormat::Unknown,
        title: None,
        height: None,
        yt_id: None,
        next_url: Some(href.to_string()),
        is_external: false,
        request_headers: None,
        label: None,
        bytes: None,
        meta_extractor_id: None,
    })
}
